src/placeos-frontend-loader/api/remote/github.cr

```diff
@@ -0,0 +1,187 @@
+          HTTP::Client.get(redirect_response.headers["location"]) do |response|
+            File.write(dest, response.body_io)
+          end
+        end
+        File.new(dest)
+      rescue ex : File::Error | HTTP::Server::ClientError
+        Log.error(exception: ex) { "Could not download file at URL: #{ex.message}" }
+      end
+
+      def extract_file(tar_name, dest_path)
+        raise File::NotFoundError.new(message: "File #{tar_name} does not exist", file: tar_name) unless File.exists?(Path.new(tar_name))
+        if !Dir.exists?(Path.new(["./", dest_path]))
+          File.open(tar_name) do |file|
+            begin
+              Compress::Gzip::Reader.open(file) do |gzip|
+                Crystar::Reader.open(gzip) do |tar|
+                  tar.each_entry do |entry|
+                    next if entry.file_info.directory?
+                    parts = Path.new(entry.name).parts
+                    parts = parts.last(parts.size > 1 ? parts.size - 1 : 0)
+                    next if parts.size == 0
+                    file_path = Path.new([dest_path] + parts)
+                    Dir.mkdir_p(file_path.dirname) unless Dir.exists?(file_path.dirname)
+                    File.write(file_path, entry.io, perm: entry.file_info.permissions)
+                  end
+                end
+              end
+            rescue ex : File::Error | Compress::Gzip::Error
+              Log.error(exception: ex) { "Could not unzip tar" }
+            end
+          end
+        end
+        File.delete(tar_name)
+      end
+
+      private def get_hashes(repo_url : String)
+        stdout = IO::Memory.new
+        Process.new("git", ["ls-remote", repo_url], output: stdout).wait
+        output = stdout.to_s.split('\n')
+        ref_hash = Hash(String, String).new
+        output.each do |ref|
+          next if ref.empty?
+          split = ref.partition('\t')
+          ref_hash[split[2]] = split[0]
+        end
+        ref_hash
+      end
+
+      private def get_hash_head(repo_url : String)
+        ref_hash = get_hashes(repo_url)
+        ref_hash.has_key?("HEAD") ? ref_hash["HEAD"] : ref_hash.first_key?
+      end
+
+      private def get_hash_by_branch(repo_url : String, branch : String)
+        ref_hash = get_hashes(repo_url)
+        raise Exception.new("Branch #{branch} does not exist in repo") unless ref_hash.has_key?("refs/heads/#{branch}")
+        ref_hash["refs/heads/#{branch}"]
+      end
+
+      private def save_metadata(path : Path, hash : String, repository_uri : String, branch : String)
+        hash_path = Path.new([path, "current_hash.txt"])
+        File.write(hash_path, hash)
+        repo_path = Path.new([path, "current_repo.txt"])
+        repo = repository_uri.partition(".com/")[2]
+        File.write(repo_path, repo)
+        branch_path = Path.new([path, "current_branch.txt"])
+        File.write(branch_path, branch)
```

Comment:
I get the need to do this, perhaps there's a slightly neater way of doing this?
I found this library that is a [disk persisted hash](https://github.com/mehmetc/hash_file), could be a good temporary solution? 
caveat, looks a bit old and may need a refresh (hopefully not too big of a deal).